### PR TITLE
[WIP] foundation for building with contrib + sockets library - main repo part I

### PR DESCRIPTION
**Pull request by maxgolov**
**Changes**

PR implements the necessary plumbing for `fluentd` / `fluentbit` exporters that live in contrib repo. Since we already have a library for basic cross-plat sockets support (used by HTTP server), I refactored that library to support TCP, UDP and Unix Domain sockets. Subsequently, `fluentd` exporter uses that library to talk to end-point over TCP, UDP or Unix Domain.

To explain the structure:
- `/ext/include/opentelemetry/ext/net/common` - network library, socket client, socket server. Reusable, (still) used by HTTP server. Now with more sockets support, i.e. in addition to TCP (that was already there) - I added UDP and Unix Domain.
- `/ext/test/sockets` - cross-plat tests for the sockets.
- `CMakeLists.txt` - that allows to build with contrib specified via local directory, or by `FetchContent` from contrib repo.

Subsequently, contrib repo part II relies on some generic / general purpose code that lives in the main, e.g. in order to avoid reimplementing the same socket library again. The code is Apache License 2.0, all copyright OpenTelemetry Authors. No need to borrow any alternate BSD or Boost -licensed network library.

@lalitb - corresponding PR for the `fluentd` exporter is in the contrib repo is here:
https://github.com/open-telemetry/opentelemetry-cpp-contrib/pull/42

### CODE REVIEW

File: CMakeSettings.json [hunk 1]
```diff
@@ -121,6 +121,86 @@
           "type": "BOOL"
         }
       ]
-    }
-  ]
+        },
```
Comment (maxgolov) [inline CMakeSettings.json:124]: I will remove this. This is just an illustration how to build with contrib from Visual Studio. I'll move this into separate document.

File: exporters/etw/test/etw_tracer_test.cc [hunk 2]
```diff
@@ -14,6 +14,8 @@ using namespace OPENTELEMETRY_NAMESPACE;
 
 using namespace opentelemetry::exporter::etw;
```
Comment (maxgolov) [inline exporters/etw/test/etw_tracer_test.cc:15]: I will revert this, as I submitted these changes in a separate PR #882

**Comment by lalitb** [comment]
I would review it more thoroughly once this is ready for review, the general comments I have for now are:

- `socket library` - Very nicely done and this would be definitely useful for applications requiring cross-platform network libraries, we need to discuss whether the `main` repo is the right place for this. While I see the possibility of using the socket library internally for Jaeger thrift implementation (  cc @ThomsonTan  ), still not a convincing reason for exposing it as api/header to external systems. As an example, we do have an internal HTTP client and (test) server implementation, but this can't be used from external systems. The `main` repo should provide components for opentelemetry-api, opentelemetry-sdk, various core exporters, and not more than that. It can be in a separate repo so that both fluentd and thrift can use from there, or else let's keep it within fluentd exporter for now?

-  `For the building with Contrib repo` - This has been discussed in community meetings and in issue #713 with some good comments from @eyjohn, would be good to have wider feedback from @open-telemetry/cpp-approvers and @open-telemetry/cpp-maintainers, as my thoughts may be biased on my general perception of the build configuration :)

**Comment by lalitb** [comment]
Closing this PR. Plan is to consolidate these changes with fluentd exporter in the `contrib` repo.

**Pull request closed without merging**